plugins/chat/assets/javascripts/discourse/services/chat-api.js

```diff
@@ -0,0 +1,248 @@
+import Service, { inject as service } from "@ember/service";
+import { ajax } from "discourse/lib/ajax";
+import UserChatChannelMembership from "discourse/plugins/chat/discourse/models/user-chat-channel-membership";
+import { tracked } from "@glimmer/tracking";
+import { bind } from "discourse-common/utils/decorators";
+import { Promise } from "rsvp";
+
+class Collection {
+  @tracked items = [];
+  @tracked meta = {};
+  @tracked loading = false;
+
+  constructor(resourceURL, handler) {
+    this._resourceURL = resourceURL;
+    this._handler = handler;
+    this._fetchedAll = false;
+  }
+
+  get loadMoreURL() {
+    return this.meta.load_more_url;
+  }
+
+  get totalRows() {
+    return this.meta.total_rows;
+  }
+
+  get length() {
+    return this.items.length;
+  }
+
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
+  [Symbol.iterator]() {
+    let index = 0;
+
+    return {
+      next: () => {
+        if (index < this.items.length) {
+          return { value: this.items[index++], done: false };
+        } else {
+          return { done: true };
+        }
+      },
+    };
+  }
+
+  @bind
+  load(params = {}) {
+    this._fetchedAll = false;
+
+    if (this.loading) {
+      return;
+    }
+
+    this.loading = true;
+
+    const queryString = Object.keys(params)
+      .map((key) => {
+        if (params[key]) {
+          return `${encodeURIComponent(key)}=${encodeURIComponent(
+            params[key]
+          )}`;
+        }
+      })
+      .filter(Boolean)
+      .join("&");
```

Comment:
IMO it's cleaner to use `URLSearchParams` here.

Also, I note you're using `.filter(Boolean)`. That means that literal `false` values will be filtered out. I imagine the intention was to only ignore `undefined` values?

```suggestion
    const filteredQueryParams = Object.entries(params).filter(
      ([, v]) => v !== undefined
    );
    const queryString = new URLSearchParams(filteredQueryParams).toString();
```